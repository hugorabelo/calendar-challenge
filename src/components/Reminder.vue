<template>
  <div>
    <div
      @click="viewDetails()"
      class="reminder-description text-sm md:text-base text-white rounded cursor-pointer mb-1"
      :class="currentReminder.color"
    >
      {{ currentReminder.description }}
    </div>
    <div v-if="showDetails" class="fixed inset-0 bg-gray-600 bg-opacity-50 overflow-y-auto h-full w-full" id="my-modal">
      <div class="relative top-20 mx-auto w-1/4 shadow-lg rounded-md bg-white">
        <div class="p-4 divide-y">
          <div class="text-right">
            <button
              v-if="!isEditing"
              id="btn-edit-reminder"
              class="rounded-full h-10 w-10 hover:bg-gray-200 text-gray-500"
              @click="editReminder()"
            >
              <font-awesome-icon icon="fa-solid fa-edit" />
            </button>
            <button
              v-if="!isEditing"
              id="btn-delete-reminder"
              class="rounded-full h-10 w-10 hover:bg-gray-200 text-gray-500"
              @click="deleteReminder()"
            >
              <font-awesome-icon icon="fa-solid fa-trash" />
            </button>
            <button
              id="btn-close-dialog"
              class="rounded-full h-10 w-10 hover:bg-gray-200 text-gray-500"
              @click="closeDetails()"
            >
              <font-awesome-icon icon="fa-solid fa-xmark" />
            </button>
          </div>
          <div v-if="isEditing" class="mt-1 pt-3">
            <reminder-form :editing-reminder="currentReminder" @save="saveEdit" @cancel="cancelEdit"></reminder-form>
          </div>
          <div v-else class="grid grid-cols-3 mt-1 pt-2">
            <div class="col-span-2 text-left">
              <div class="flex">
                <h3 class="text-lg leading-6 font-medium text-gray-900 py-1">
                  {{ currentReminder.description }}
                </h3>
              </div>
              <div class="mt-3">
                <p class="text-sm text-gray-900">
                  {{
                    `${getFullFormattedDate(currentReminder.dateTime)} - ${getFormattedTime(currentReminder.dateTime)}`
                  }}
                </p>
                <p class="text-sm text-gray-900">
                  {{ currentReminder.city }}
                </p>
              </div>
            </div>
            <div class="flex mt-3 items-center">
              <div class="flex flex-col float-right mx-auto mr-0">
                <div
                  class="flex w-16 h-16 rounded-full bg-opacity-40 justify-center items-center mx-auto"
                  :class="currentReminder.color"
                >
                  <img
                    v-if="currentReminder.weatherIcon"
                    :src="`http://openweathermap.org/img/wn/${currentReminder.weatherIcon}@2x.png`"
                  />
                  <font-awesome-icon v-else icon="fa-solid fa-ban" class="text-gray-500" size="2x" />
                </div>
                <div class="text-sm text-gray-900">{{ currentReminder.weatherForecast }}</div>
              </div>
            </div>
          </div>
        </div>
        <div
          v-if="currentReminder.color && !isEditing"
          class="h-7 rounded-lg border"
          :class="currentReminder.color"
        ></div>
      </div>
    </div>
  </div>
</template>

<script>
import ReminderForm from './ReminderForm.vue'
import dateTimeMixins from '@/mixins/dateTimeMixins'
import * as reminderApi from '@/services/reminders'

export default {
  name: 'Reminder',
  components: {
    ReminderForm,
  },
  mixins: [dateTimeMixins],
  props: {
    reminderData: {
      type: Object,
      default: function () {
        return {
          id: null,
          description: null,
          dateTime: null,
          city: null,
          color: null,
          weatherForecast: null,
          weatherIcon: null,
        }
      },
    },
  },
  data() {
    return {
      currentReminder: {
        ...this.reminderData,
        dateTime: this.getObjectDateTime(this.reminderData.dateTime),
      },
      showDetails: false,
      isEditing: false,
    }
  },
  methods: {
    async viewDetails() {
      this.isEditing = false
      this.showDetails = true
    },
    closeDetails() {
      this.showDetails = false
    },
    editReminder() {
      this.isEditing = true
    },
    cancelEdit() {
      this.isEditing = false
    },
    async saveEdit(newReminder) {
      this.currentReminder = newReminder
      this.isEditing = false
    },
    async deleteReminder() {
      await reminderApi.deleteReminder(this.currentReminder)
      this.showDetails = false
      this.$destroy()
    },
  },
}
</script>

<style></style>
