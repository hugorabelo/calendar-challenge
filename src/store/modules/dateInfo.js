import { DateTime } from 'luxon'

const state = {
  selectedMonth: DateTime.now().startOf('month'),
}
const getters = {
  selectedMonth: (state) => state.selectedMonth,
}
const mutations = {
  setSelectedMonth(state, data) {
    state.selectedMonth = data
  },
}
const actions = {
  initSelectedMonth({ commit }) {
    commit('setSelectedMonth', DateTime.now().startOf('month'))
  },
  goToNextMonth({ commit, state, dispatch }) {
    if (!state.selectedMonth) {
      dispatch('initSelectedMonth')
    }
    let nextMonth = state.selectedMonth.plus({ months: 1 })
    commit('setSelectedMonth', nextMonth)
  },
  goToPreviousMonth({ commit, state, dispatch }) {
    if (!state.selectedMonth) {
      dispatch('initSelectedMonth')
    }
    let previousMonth = state.selectedMonth.minus({ months: 1 })
    commit('setSelectedMonth', previousMonth)
  },
}

export default {
  state,
  getters,
  actions,
  mutations,
}
