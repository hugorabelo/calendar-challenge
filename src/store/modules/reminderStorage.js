const state = {
  reminders: localStorage.getItem('reminders') != null ? JSON.parse(localStorage.getItem('reminders')) : [],
}
const getters = {
  reminders: (state) => state.reminders,
}
const mutations = {
  setReminders(state, data) {
    state.reminders = data
    localStorage.setItem('reminders', JSON.stringify(state.reminders))
  },
}
const actions = {
  addReminder({ commit, state }, data) {
    if (state.reminders.filter((rem) => rem.id == data.id).length == 0) {
      state.reminders.push(data)
    }
    commit('setReminders', state.reminders)
  },
  updateReminder({ commit, state }, data) {
    for (var i = 0; i < state.reminders.length; i++) {
      if (state.reminders[i].id == data.id) {
        state.reminders[i] = data
      }
    }
    commit('setReminders', state.reminders)
  },
  deleteReminder({ commit, state }, data) {
    for (var i = 0; i < state.reminders.length; i++) {
      if (state.reminders[i].id == data.id) {
        state.reminders.splice(i, 1)
      }
    }
    commit('setReminders', state.reminders)
  },
  deleteRemindersGroup({ dispatch }, data) {
    for (var i = 0; i < data.length; i++) {
      dispatch('deleteReminder', { id: data[i].id })
    }
  },
}

export default {
  state,
  getters,
  actions,
  mutations,
}
