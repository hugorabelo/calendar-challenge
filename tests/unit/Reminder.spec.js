import { mount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import Reminder from '@/components/Reminder'
import { DateTime } from 'luxon'
import dateInfo from '@/store/modules/dateInfo'
import reminderStorage from '@/store/modules/reminderStorage'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
import { library } from '@fortawesome/fontawesome-svg-core'
import { faBan, faEdit, faTrash, faXmark } from '@fortawesome/free-solid-svg-icons'

const localVue = createLocalVue()
localVue.use(Vuex)
library.add(faBan, faEdit, faTrash, faXmark)
localVue.component('font-awesome-icon', FontAwesomeIcon)

describe('Reminder', () => {
  const spy = jest.fn()

  const reminder = {
    id: '123-456',
    description: 'Event 01',
    dateTime: DateTime.now(),
    city: 'Los Angeles',
    color: 'bg-red-600',
    weatherForecast: 'light rain',
    weatherIcon: '10n',
  }

  const build = (options) => {
    return mount(Reminder, {
      localVue,
      store: new Vuex.Store({
        modules: {
          dateInfo: { ...dateInfo },
          reminderStorage: { ...reminderStorage },
        },
      }),
      ...options,
      destroyed() {
        spy()
      },
    })
  }

  test('Is a Vue instance', () => {
    const { vm } = build({
      propsData: {
        reminderData: reminder,
      },
    })
    expect(vm).toBeTruthy()
  })

  test('Click Dialog Buttons', async () => {
    const wrapper = build()

    expect(wrapper.html()).toMatchSnapshot()

    let labelReminder = wrapper.findComponent('.reminder-description')
    await labelReminder.trigger('click')
    expect(wrapper.vm.showDetails).toBeTruthy()

    let btnClose = wrapper.findComponent('#btn-close-dialog')
    await btnClose.trigger('click')
    expect(wrapper.vm.showDetails).toBeFalsy()

    await labelReminder.trigger('click')

    let btnEdit = wrapper.findComponent('#btn-edit-reminder')
    await btnEdit.trigger('click')
    expect(wrapper.vm.isEditing).toBeTruthy()
  })

  test('Delete Reminder', async () => {
    const wrapper = build()

    const store = new Vuex.Store({
      modules: {
        reminderStorage: reminderStorage,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()

    let labelReminder = wrapper.findComponent('.reminder-description')
    await labelReminder.trigger('click')
    expect(wrapper.vm.showDetails).toBeTruthy()

    store.state.reminderStorage.reminders = [reminder]

    expect(store.state.reminderStorage.reminders.length).toBe(1)

    let btnDelete = wrapper.findComponent('#btn-delete-reminder')
    await btnDelete.trigger('click')
    expect(spy).toHaveBeenCalled()
  })
})
