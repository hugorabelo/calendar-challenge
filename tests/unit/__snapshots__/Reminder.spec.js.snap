// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reminder Click Dialog Buttons 1`] = `
<div>
  <div class="reminder-description text-sm md:text-base text-white rounded cursor-pointer mb-1">

  </div>
  <!---->
</div>
`;

exports[`Reminder Delete Reminder 1`] = `
<div>
  <div class="reminder-description text-sm md:text-base text-white rounded cursor-pointer mb-1">

  </div>
  <!---->
</div>
`;
